listenbrainz_spark/troi/periodic_jams.py

```diff
@@ -0,0 +1,53 @@
+from datetime import datetime, timedelta
+
+from more_itertools import chunked
+
+from listenbrainz_spark.path import RECORDING_FEEDBACK_DATAFRAME, RAW_RECOMMENDATIONS
+from listenbrainz_spark.stats import run_query
+
+DAYS_OF_RECENT_LISTENS_TO_EXCLUDE = 60
+USERS_PER_MESSAGE = 100
+MAX_TRACKS_PER_PLAYLIST = 50
+
+
+def main(slug):
+    """ Bulk generate troi playlists for the given slug """
+    if slug == "weekly-jams":
+        max_listened_at = datetime.now() - timedelta(days=DAYS_OF_RECENT_LISTENS_TO_EXCLUDE)
+        max_listened_at = max_listened_at.replace(hour=0, minute=0, second=0, microsecond=0)
+        time_filter = f"latest_listened_at < to_timestamp('{max_listened_at}')"
+    elif slug == "weekly-exploration":
+        time_filter = "latest_listened_at IS NULL"
+    else:
+        return []
+
+    query = f"""
+        WITH recommendations AS (
+            SELECT user_id
+                 , recording_mbid
+                 , rank() OVER (PARTITION BY user_id ORDER BY score DESC) AS position
```

Comment:
the recommendation tracks should be sorted randomly, not by order. In troi the playlist gets shuffled before being truncated, but that is hidden in the playlist generator.

Reply:
Shuffle before truncating in all cases?

Reply:
I actually went back and looked at exactly what troi does and its the other way around. Sort by ranking, truncate THEN shuffle. Not how I had originally intended, but it works well, so let's continue to do that.

Reply:
Fixed.